### ISSUE TITLE
Support inline commenting

### ISSUE THREAD

**Issue by noahgitsham** [neovim/neovim#29834]
### Problem

The builtin commenting feature added in 0.10 is great, though I often find myself looking to comment out select portions of a line, rather than the line in its entirety.

It would be nice if inline comments could be inserted and removed natively for languages that support them.

### Expected behavior

`gC` seems like a perfect candidate for this.

It could also be implemented as a text object to select, delete inline comments under the cursor etc.

## Examples:
 (with C++ style comments, `|` represents the cursor over the character to the right)

### Inner word
```c++
some te|xt bla bla bla
```
Type `gCiw`
```c++
some |/*text*/ bla bla bla
```

### To end of line
```c++
some te|xt bla bla bla
```
Type `gC$`
```c++
some te|/*xt bla bla bla*/
```

### From selection
```c++
some te|xt bla bla bla
```
Type `veegC`
```c++
some te|/*xt bla*/ bla bla
```

### Multiline
```c++
some multi|line
textbla bla bla bla
```
Type `vjgC`/`gCj`
```c++
some multi|/*line
textbla bla*/ bla bla
```

### Multiline visual block
```c++
some multi|line
textbla bla bla bla
```
Type `<ctrl-v>jlllgC`
```c++
some multi|/*line*/
textbla bl/*a bl*/a bla
```

### Uncommenting
```c++
/*some te|xt bla bla bla*/
```
Type `gCiw`
```c++
/*some */|text/* bla bla bla*/
```

**Comment by clason** [comment]
Sorry, that is out of scope for the builtin feature; inline comments are way too niche and better left to a plugin.

**Comment by justinmk** [comment]
Related: https://github.com/neovim/neovim/issues/29318

**Comment by Konfekt** [comment]
> way too niche 

What's niche ? Just grepping for `'\S\s+"\s*[a-zA-Z0-9 ]+$'` inside nvim's runtime yields more than 10k hits

**Comment by clason** [comment]
Don't forget that this includes thousands of files inherited from Vim, which we don't touch as a matter of principle.

**Comment by Konfekt** [comment]
I have to correct and be more precise `grep -E '\S\s+"\s*[a-z ]+$' */*.vim | wc -l ` yields 271 matches. Still, this happens even outside Vim, for example in shell scripts, say `xdg-settings` by Google (just an arbitrary example that came to mind).

**Comment by clason** [comment]
Sorry, that may have been phrased too glibly. Obviously inline commenting is a thing; we're just not convinced that the added complexity is worth it _for the bundled module_. The point of including is was to have basic `gcc` work, _not_ to completely replace all the commenting plugins out there. If you have more complex needs, you can (and should) install one of the many commenting plugins, which still work fine.

**Comment by Konfekt** [comment]
Fair enough. It just seemed more in sync with the other upper-case mappings `C/Y/D` operating on the rest of the line to let `gC` do so as well.

**Comment by Konfekt** [comment]
So why not set `gC` in `ftplugin` whenever `commentstring` is applicable in-line? That's not too rare.

**Comment by Konfekt** [comment]
Maybe we continue discussion https://github.com/vim/vim/issues/15727

**Comment by clason** [comment]
Even with this additional option, this introduces a much larger API surface (what about uncommenting?) And I'm not sure the analogy with `C/Y/D` holds much water. I understand why it can be desirable, but I don't see a strong argument for why it _must_ be builtin. (It can, but the price seems too high to me.)

**Comment by Konfekt** [comment]
Not sure what holding water here refers to, but as a newcomer one could draw an analogy that the upper case version of a built-in operator usually operates to the rest of the line. Adding a `nmap <buffer> gC gc$`(if ever `gc$` understood that it meant inserting the comment string at the cursor position) in suitable ftplugins like python, sh, ... is not too much an effort; of course, many more complicated things would be down, but more than 80% of the common uses cases are covered.

Maybe ftplugin, though the most practical solution, is not strictly necessary, but only checking that nothing follows `%s` in `&commentstring`.

**Comment by clason** [comment]
>  upper case version of a built-in operator usually operates to the rest of the line.

That's what I'm referring to -- there is no such intended convention about _any_ operator; just these special three. 

And remember that the point of a line in the sand is not to guard against a single, potentially reasonable-sounding, request: it's to guard against the _totality_ of all such requests. So the bar for extending the builtin, intentionally _very_ minimal, commenting feature is high.
